### PR TITLE
Issue #4944: Extend Checkstyle GUI tool to work with Xpath

### PR DESCRIPTION
**Pull request by developerhb**
Issue #4944: Extend Checkstyle GUI tool to work with Xpath

This PR is to Find Node by Xpath. 

### Preview

**Correct Xpath Query:**

**Before**

image

**After**

image

**Incorrect Xpath Query:**

**Before:**

image

**After:**

image

### CODE REVIEW

**Comment by developerhb** [comment]
@romani 

I have made use of Stack in the code. Also, I have used recursion so my function is having multiple return statements for multiple conditions.

I am getting following errors in mvn clean verify

```
[INFO] --- maven-checkstyle-plugin:3.1.1:check (sevntu-checkstyle-check) @ checkstyle ---
[INFO] There are 3 errors reported by Checkstyle 8.29 with F:\College\checkstyle/config/checkstyle_sevntu_checks.xml ruleset.
[ERROR] src\main\java\com\puppycrawl\tools\checkstyle\gui\TreeTable.java:[31,1] (extension) ForbidCertainImports: Import 'java.util.Stack' should not match 'java\.util\.Stack|java\.util\.Vector|             com\.google\.collect\.Iterables|com\.google\.common\.annotations\.Beta|             com\.google\.common\.base\.Predicate|com\.google\.common\.base\.String||             com\.google\.common\.base\.Function|com\.google\.common\.base\.Supplier|             com\.google\.common\.base\.Charsets|com\.google\.common\.base\.MoreObjects|             com\.google\.common\.base\.Equivalence|com\.google\.common\.base\.Preconditions|             com\.google\.common\.base\.Optional|com\.google\.common\.io\.CharSource|             com\.google\.common\.primitives.*|com\.google\.common\.collect\.Sets|             com\.google\.common\.collect\.Queues|com\.google\.common\.collect\.Lists|             com\.google\.common\.annotations\.GwtCompatible|com\.google\.common\.io\.Files|             com\.google\.common\.collect\.TreeTraverser' pattern, it is forbidden.
[ERROR] src\main\java\com\puppycrawl\tools\checkstyle\gui\TreeTable.java:[202,21] (extension) ReturnCountExtended: Return count for 'search' method is 2 (max allowed is 1).
[ERROR] src\main\java\com\puppycrawl\tools\checkstyle\gui\TreeTable.java:[230,38] (extension) ForbidInstantiation: Instantiation of 'Stack' is not allowed.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
```
Note:
I have added <allow class="java.util.Stack" /> in import-control.xml.

Help me with this.

**Comment by romani** [comment]
@developerhb ,

please read https://docs.oracle.com/javase/7/docs/api/java/util/Stack.html
please use `Deque<Integer> stack = new ArrayDeque<Integer>();`

File: src/main/java/com/puppycrawl/tools/checkstyle/gui/TreeTable.java [hunk 1]
```diff
@@ -190,6 +191,63 @@ private void setColumnsInitialWidth() {
...
+            {
+                if(search(child, xpath, nodes)) {
+                    nodes.push(root);
+                    return true;
+                }
+                child = child.getNextSibling();
+            }
+        }
+        return false;
+    }
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/gui/TreeTable.java:220]: something like:

```
    private boolean search(DetailAST root, String xpath, Stack<DetailAST> nodes) {
       boolean result = false;
        if(xpath.equals(XpathQueryGenerator.generateXpathQuery(root))) {
                nodes.push(root);
                result = false;
        } else {
        if(root.hasChildren()) {
            DetailAST child = root.getFirstChild();    
            while(child!=null)
            {
                if(search(child, xpath, nodes)) {
                    nodes.push(root);
                    result = true;
                    break;
                }
                child = child.getNextSibling();
            }
        }

        }

        return result;
    }
```

File: config/import-control.xml [hunk 2]
```diff
@@ -159,8 +159,8 @@
     <!-- until https://github.com/checkstyle/checkstyle/issues/3095 -->
     <allow pkg="antlr"/>
     <allow class="com.puppycrawl.tools.checkstyle.DetailAstImpl"/>
-    <allow class="java.nio.charset.StandardCharsets" />
```
Comment (romani) [inline config/import-control.xml:162]: expalin this update.

if it is unrelated to your changes but non-required config - please send removal in separarte PR.

Comment (developerhb) [reply config/import-control.xml:162]: @romani I added `<allow class="java.nio.charset.StandardCharsets" />` in my previous PR https://github.com/checkstyle/checkstyle/pull/8317. Later on I removed the code which required this import, but forgot to remove from `import-control.xml`.

Should I keep it in this PR or create a seperate PR for this?

Comment (romani) [reply config/import-control.xml:162]: Keep in this PR but please move this change to separate commit, commit should have "minor:" prefix. This PR Will have 2 commits

Comment (developerhb) [reply config/import-control.xml:162]: Okay @romani

File: src/main/java/com/puppycrawl/tools/checkstyle/gui/TreeTable.java (hunk 1 above)
Comment (developerhb) [reply src/main/java/com/puppycrawl/tools/checkstyle/gui/TreeTable.java:223]: Done

**Comment by developerhb** [comment]
@romani this PR is ready from my side

File: config/import-control.xml (hunk 2 above)
Comment (romani) [reply config/import-control.xml:162]: Please swap commits in order, functional fix should be last commit in your branch

File: src/main/java/com/puppycrawl/tools/checkstyle/gui/TreeTable.java [hunk 3]
```diff
@@ -190,6 +192,64 @@ private void setColumnsInitialWidth() {
...
+                    path = path.pathByAddingChild(nodes.pop());
+                    if (!tree.isExpanded(path)) {
+                        tree.expandPath(path);
+                    }
+                    tree.setSelectionPath(path);
+                    makeCodeSelection();
+                }
+            }
+            else {
+                xpathEditor.setText(xpathEditor.getText() + "- wrong xpath query");
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/gui/TreeTable.java:245]: `\n^ wrong xpath query`

Comment (developerhb) [reply src/main/java/com/puppycrawl/tools/checkstyle/gui/TreeTable.java]: @romani done

File: config/import-control.xml (hunk 2 above)
Comment (developerhb) [reply config/import-control.xml:162]: done @romani

**Comment by developerhb** [comment]
> I would be awesome to avoid recursion but it is ok, it is unlikely that somebody will use this GUI for huge files.

Yes. I also tried to avoid recursion but was not able to find other way.

File: src/main/java/com/puppycrawl/tools/checkstyle/gui/TreeTable.java [hunk 4]
```diff
@@ -190,6 +192,64 @@ private void setColumnsInitialWidth() {
...
+     * @return {@code boolean} node found or not
+     */
+    private static boolean search(DetailAST root, String xpath, Deque<DetailAST> nodes) {
+        boolean result = false;
+        if (xpath.equals(XpathQueryGenerator.generateXpathQuery(root))) {
+            nodes.push(root);
+            result = true;
+        }
+        else {
+            if (root.hasChildren()) {
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/gui/TreeTable.java:210]: This `if` is redundant. When the root has no children, `root.getFirstChild()` will return `null` and the entire loop will be skipped.

Comment (developerhb) [reply src/main/java/com/puppycrawl/tools/checkstyle/gui/TreeTable.java]: @pbludov Done

**Comment by developerhb** [comment]
This PR is merged. Thank you @romani and @pbludov for guiding and reviewing this PR

**Pull request merged**